### PR TITLE
[vrotsc,vrotsc-annotations] (#336) Add Workflow Canvas Item for User Interaction Component

### PR DESCRIPTION
**Pull request by akantchev**
### User Interaction Workflow Item

The decorator is used to specify an user interaction workflow item.

#### Supported Parameters

- `target` - The name of the target to that user interaction workflow item is connected to. You can specify another user interaction workflow item as a target, thus chaining multiple user interaction components.

In order to bind inputs and outputs, you do it with the `@In` and `@Out` decorators. 

#### Inputs

If you need to specify certain access limitation for the user interaction component you can specify them with `@In` decorators.
Those inputs are optional.

- `security_assignees` (type `Array/LdapUser`) - Any user from this array of users will be authorized to fill in this form.
- `security_assignee_groups` (type `Array/LdapGroup`) - Any user member of any of the groups will be authorized to fill in this form.
- `security_group` (type  `LdapGroup`) - Any user member of this group will be authorized to fill in this form.
- `timeout_date` (type  `Date`) - If not null, this input item will wait until date and will continue workflow execution.

Note that those parameters should match also the input parameters of the workflow.

##### Known Limitations for the Input Parameters

The names of the variables in the additional method decorators should be as following:
- `security_assignees` - for the security assignees parameter.
- `security_assignee_groups` - for the security assignee group parameter.
- `security_group` - for the security group parameter.
- `timeout_date` - for the timeout date parameter.

#### Outputs

You can specify multiple output variables that would hold the answer of the user interaction components.

#### Example Workflow

```ts
import {
    Workflow,
    In,
    Item,
    RootItem,
    UserInteractionItem,
} from "vrotsc-annotations";

@Workflow({
    name: "User Interaction",
    path: "VMware/PSCoE",
    description: "Adding user interaction parameters",
})
export class UserInteractionWorkflow {

    @RootItem({ target: "userInteraction1Enter", exception: "" })
    public start() {
        System.log("Starting workflow");
    }

    @UserInteractionItem({
        target: "userInteraction2Enter",
    })
    public userInteraction1Enter() {
        System.log("User interaction component 1 activation");
    }

    @UserInteractionItem({
        target: "userInteractionExit",
    })
    public userInteraction2Enter(
        @In security_assignees: LdapUser[],
        @In security_assignee_groups: LdapGroup[],
        @In security_group: LdapGroup,
        @In timeout_date?: Date,
        @In userInteractionAnswer?: string
    ) {
        System.log("User interaction component 2 activation");
    }

    @Item({ target: "end" })
    public userInteractionExit(@In userInteractionAnswer: string) {
        System.log("User Interaction exit");
    }
```
User_Interaction_Canvas_Item_Workflow

### Checklist

- [x] I have added relevant error handling and logging messages to help troubleshooting
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation, relevant usage information (if applicable)
- [x] I have updated the PR title with affected component, related issue number and a short summary of the changes introduced
- [x] I have added labels for implementation kind (kind/*) and version type (version/*)
- [x] I have tested against live environment, if applicable
- [x] I have my changes rebased and squashed to the minimal number of relevant commits. **Notice: don't squash all commits**
- [x] I have added a descriptive commit message with a short title, including a `Fixed #XXX -` or `Closed #XXX -` prefix to auto-close the issue

### Testing

End to end tests for the user interaction component were also added.

Related issue:

#336

### CODE REVIEW

File: docs/versions/latest/Components/Archetypes/typescript/Components/Workflows.md [hunk 1]
```diff
@@ -366,3 +364,91 @@ export class HandleNetworkConfigurationBackup {
   }
 }
 ```
+
+### `@UserInteractionWorkflowItem`
+
+The decorator is used to specify an user interaction workflow item.
+
+#### Supported Parameters
```
Comment (Michaelpalacce) [inline docs/versions/latest/Components/Archetypes/typescript/Components/Workflows.md:372]: Maybe we should specify how to configure the form? It's following:

{WORKFLOW_NAME}_{USER_INTERACTION_ACTION}.form.json, correct?

File: typescript/vrotsc/src/compiler/transformer/fileTransformers/workflow/decorators/decisionItemDecoratorStrategy.ts [hunk 2]
```diff
@@ -40,19 +39,21 @@ import { formatPosition } from "../helpers/formatPosition";
    <position y="40.0" x="380.0"/>
  </workflow-item>
  */
-export default class DecisionItemDecoratorStrategy implements CanvasItemDecoratorStrategy {
-
-	constructor(private readonly sourceFilePrinter: SourceFilePrinter = new WrapperSourceFilePrinter()) { }
+export default class DecisionItemDecoratorStrategy extends BaseItemDecoratorStrategy {
```
Comment (Michaelpalacce) [inline typescript/vrotsc/src/compiler/transformer/fileTransformers/workflow/decorators/decisionItemDecoratorStrategy.ts:42]: I disagree to extending this logic like this, we don't gain anything but complexity..

File: typescript/vrotsc/src/compiler/transformer/fileTransformers/workflow/decorators/decisionItemDecoratorStrategy.ts [hunk 3]
```diff
@@ -81,25 +82,14 @@ export default class DecisionItemDecoratorStrategy implements CanvasItemDecorato
...
-				findTargetItem(itemInfo.target, pos, itemInfo),
-				findTargetItem((itemInfo.canvasItemPolymorphicBag as CanvasItemPolymorphicBagForDecision).else, pos, itemInfo)
-			],
-			offset: [0, -10]
-		};
-
-		if (itemInfo.canvasItemPolymorphicBag.exception) {
-			node.targets.push(findTargetItem(itemInfo.canvasItemPolymorphicBag.exception, pos, itemInfo));
-		}
+	public getGraphNode(itemInfo: WorkflowItemDescriptor, pos: number): GraphNode {
```
Comment (Michaelpalacce) [inline typescript/vrotsc/src/compiler/transformer/fileTransformers/workflow/decorators/decisionItemDecoratorStrategy.ts:85]: Again, bad change imho

File: typescript/vrotsc/src/compiler/transformer/fileTransformers/workflow/decorators/base/baseItemDecoratorStrategy.ts [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        }
+
+        return node;
+    }
+
+    protected buildParameterBindings(itemInfo: WorkflowItemDescriptor, bindingsType: InputOutputBindings): string {
+        return buildItemParameterBindings(itemInfo, bindingsType);
+    }
+
+    protected findTargetItem(target: any, pos: number, item: WorkflowItemDescriptor): string {
```
Comment (Michaelpalacce) [inline typescript/vrotsc/src/compiler/transformer/fileTransformers/workflow/decorators/base/baseItemDecoratorStrategy.ts:141]: why

File: typescript/vrotsc/src/compiler/transformer/fileTransformers/workflow/decorators/base/baseItemDecoratorStrategy.ts [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        return findTargetItem(target, pos, item);
+    }
+
+    protected formatItemPosition(position: [number, number]): string {
```
Comment (Michaelpalacce) [inline typescript/vrotsc/src/compiler/transformer/fileTransformers/workflow/decorators/base/baseItemDecoratorStrategy.ts:145]: why?

**Review by Michaelpalacce — DISMISSED** [review]
I don't agree with this approach. I am not going to approve these changes. If someone else wants to do so, feel free, I will remove my review then

File: typescript/vrotsc/src/compiler/transformer/fileTransformers/workflow/decorators/base/baseItemDecoratorStrategy.ts (hunk 4 above)
Comment (akantchev) [reply typescript/vrotsc/src/compiler/transformer/fileTransformers/workflow/decorators/base/baseItemDecoratorStrategy.ts:141]: why what?

File: typescript/vrotsc/src/compiler/transformer/fileTransformers/workflow/decorators/base/baseItemDecoratorStrategy.ts (hunk 5 above)
Comment (akantchev) [reply typescript/vrotsc/src/compiler/transformer/fileTransformers/workflow/decorators/base/baseItemDecoratorStrategy.ts:145]: why what?

**Review by joroaf — APPROVED** [review]

**Comment by Michaelpalacce** [comment]
I can't believe we merged this 😮

We are using the "Strategy Pattern" for this. The whole point it was done this way was so we can avoid inheritance... https://www.geeksforgeeks.org/favoring-composition-over-inheritance-in-java-with-examples/

Now we've created an entirely useless abstract parent, that easily could have been a function if we were so worried about duplication. 

I am not saying to never use inheritance, but doing so in a Strategy pattern is... atrocious. 

Not to mention the wrappers that were added to the parent class. The way it was done, completely breaks SRP.

Let's introduce more complexity for the sake of complexity in this "overly simple project", it boggles my mind.

**Pull request merged**